Fix v8_enable_handle_zapping for 4.5

skomski opened the pull request:
upstream disabled v8_enable_handle_zapping on the debug build
and not on the release which would be correct

Correct codereview for 4.6:
https://codereview.chromium.org/1302343002

Wrong codereview for 4.5:
https://codereview.chromium.org/1229243002

Found this via this issue https://github.com/nodejs/node/issues/2721 because it only crashed now. Handle zapping in IntegerValue was exposing a bug (thanks to @indutny) but it still should be disabled for performance reasons https://code.google.com/p/chromium/issues/detail?id=318206

Our variable did not overwrite the variable in v8:
https://github.com/nodejs/node/blob/master/common.gypi#L86

indutny commented:
Why didn't our `common.gyp` override it?

bnoordhuis commented:
I had a quick look at #2721 and I don't handle zapping is the issue, it's exposing the bug.  Consider:

```c++
Maybe<int64_t> Value::IntegerValue(Local<Context> context) const {
  auto obj = Utils::OpenHandle(this);
  i::Handle<i::Object> num;
  if (obj->IsNumber()) {
    num = obj;
  } else {
    PREPARE_FOR_EXECUTION_PRIMITIVE(context, "IntegerValue", int64_t);
    has_pending_exception =
        !i::Execution::ToInteger(isolate, obj).ToHandle(&num);
    RETURN_ON_FAILED_EXECUTION_PRIMITIVE(int64_t);
  }
  return Just(num->IsSmi() ? static_cast<int64_t>(i::Smi::cast(*num)->value())
                           : static_cast<int64_t>(num->Number()));
}
```

The macros hide it but PREPARE_FOR_EXECUTION_PRIMITIVE is basically a scoped operation.  Outside the `else` block, `num` is no longer rooted.  #2722 fixes the scoping bug.

indutny commented:
@bnoordhuis yeah, zapping is not the problem. It is just causing the crash, this is why I think that #2722 is indeed a proper fix.

skomski commented:
I know this is not a fix for IntegerValue (the issue only was the cause I found this issue) but we should disable handle_zapping on release because it causes performance problems that's why the v8 team disables it for chromium https://code.google.com/p/chromium/issues/detail?id=318206

bnoordhuis commented:
Oh, I don't disagree that we should turn it off before the v4.0.0 release but it's good it's been on so far.

@indutny already mentioned it but I'll repeat it: it should be possible to toggle handle zapping from common.gypi.  If that doesn't work (and it looks like it doesn't), then it needs some further investigation.

skomski commented:
Updated. It seems the `Releases` from `features.gypi` is evaluated before `common.gypi` which results with `%` at the end of the variable into not setting the variable because it is not undefined. I removed `%` to force it.

indutny commented:
Oh that GYP order of variables thing. LGTM, if @bnoordhuis is ok with this too. @bnoordhuis PTAL

bnoordhuis commented:
LGTM but can you mention in the commit log that this disables handle zapping in release builds (or the other way around: that it was enabled until now.)

It's curious that I've never seen ZapRange() pop up during profiling.  Anyway, @indutny can you land this?  I'm off to bed.

indutny commented:
Yep, I can land this. Thanks @bnoordhuis !

kkoopa commented:
This was added to 4.5 five hours ago https://codereview.chromium.org/1326693004

rvagg commented:
landed @ 297b9ae

kkoopa commented:
@danielkhan Performance should have improved even more through this.
